File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/user/UserUpdate.java
package top.likoslupus.cellulosesz.api.user;

import org.jspecify.annotations.Nullable;

import static java.util.Objects.requireNonNull;

public record UserUpdate<T extends @Nullable Object>(
        CellUser user,
        T result
) {

    public UserUpdate {
        requireNonNull(user, "user");
    }

    public static UserUpdate<@Nullable Void> replacing(CellUser user) {
        return new UserUpdate<>(user, null);
    }

    public static <T extends @Nullable Object> UserUpdate<T> of(CellUser user, T result) {
        return new UserUpdate<>(user, result);
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/user/UserUpdate.kt
package top.likoslupus.cellulosesz.api.user

@JvmRecord
public data class UserUpdate<T>(
    public val user: CellUser,
    public val result: T
) {

    public companion object {

        @JvmStatic
        public fun replacing(user: CellUser): UserUpdate<Void?> =
            UserUpdate(user, null)

        @JvmStatic
        public fun <T> of(user: CellUser, result: T): UserUpdate<T> =
            UserUpdate(user, result)

    }

}
